File: NeetCode 150/Graphs/210. Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        #Topological Sort using BFS
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for course,pre in prerequisites:
            graph[pre].append(course)
            indegree[course] += 1
        
        res = []
        q = deque()

        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
        
        while q:
            node = q.popleft()
            res.append(node)
            for nei in graph[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)
        
        return res if len(res) == numCourses else []

File: NeedCode 150 (JAVA)/Graphs/210. Course Schedule II.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[numCourses];
        
        for(int i = 0; i<numCourses ; i++){
            graph.add(new ArrayList<>());
        }

        for(int[] p: prerequisites){
            int course = p[0];
            int prereq = p[1];

            graph.get(prereq).add(course);
            indegree[course]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for(int i =0;i<numCourses;i++){
            if(indegree[i] == 0)
                queue.offer(i);
        }
        int idx = 0;
        int[] res = new int[numCourses];

        while(!queue.isEmpty()){
            int curr = queue.poll();
            res[idx++] = curr; //idx incrementing in place

            for(int next : graph.get(curr)){
                indegree[next]--;
                if(indegree[next] == 0)
                    queue.offer(next);
            }
        }
        return idx == numCourses ? res : new int[0]; //here as well you need a new object for []
    }
}
